firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java

```diff
@@ -226,93 +349,85 @@ public void onClick(DialogInterface dialogInterface, int i) {
+                  firebaseAppDistributionTesterApiClient.fetchLatestRelease(
+                      fid, appId, apiKey, authToken);
+
+              Context context = firebaseApp.getApplicationContext();
+              long currentInstalledVersionCode = getInstalledAppVersionCode(context);
+
+              if (Long.parseLong(latestRelease.getBuildVersion()) <= currentInstalledVersionCode) {
+                // only return a release if device can upgrade to the new release without
+                // uninstalling
+                latestRelease = null;
+              }
 
-  @Override
-  public void onActivityCreated(
-      @NonNull @NotNull Activity activity, @androidx.annotation.Nullable @Nullable Bundle bundle) {
-    Log.d(TAG, "Created activity: " + activity.getClass().getName());
-    // if signinactivity is created, sign-in was succesful
-    if (currentlySigningIn && activity instanceof SignInResultActivity) {
-      currentlySigningIn = false;
-      signInTaskCompletionSource.setResult(null);
-    }
+              AppDistributionRelease finalLatestRelease = latestRelease;
+              updateOnUiThread(
+                  new Runnable() {
+                    @Override
+                    public void run() {
+                      checkForUpdateTaskCompletionSource.setResult(finalLatestRelease);
+                    }
+                  });
+            } catch (FirebaseAppDistributionException | ProtocolException e) {
+              setCheckForUpdateTaskCompletionError(
+                  new FirebaseAppDistributionException(
+                      FirebaseAppDistributionException.Status.NETWORK_FAILURE));
+            }
+          }
+        });
   }
 
-  @Override
-  public void onActivityStarted(@NonNull @NotNull Activity activity) {
-    Log.d(TAG, "Started activity: " + activity.getClass().getName());
+  private void updateOnUiThread(Runnable runnable) {
+    HandlerCompat.createAsync(Looper.getMainLooper()).post(runnable);
   }
 
-  @Override
-  public void onActivityResumed(@NonNull @NotNull Activity activity) {
-    Log.d(TAG, "Resumed activity: " + activity.getClass().getName());
-
-    // signInActivity is only opened after successful redirection from signIn flow,
-    // should not be treated as reentering the app
-    if (activity instanceof SignInResultActivity) {
-      return;
-    }
-
-    // throw error if app reentered during signin
-    if (currentlySigningIn) {
-      currentlySigningIn = false;
-      setSignInTaskCompletionError(new FirebaseAppDistributionException(AUTHENTICATION_FAILURE));
+  private long getInstalledAppVersionCode(Context context) {
+    PackageInfo pInfo = null;
+    try {
+      pInfo = context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
+    } catch (PackageManager.NameNotFoundException e) {
+      checkForUpdateTaskCompletionSource.setException(
+          new FirebaseAppDistributionException(FirebaseAppDistributionException.Status.UNKNOWN));
```

Comment:
in general, I think we should pass information about the root cause into the FirebaseAppDistributionException. Can we add e.getMessage() to the FADException?